miner/src/service_transaction_checker.rs

```diff
@@ -16,37 +16,111 @@
 		let sender = tx.sender();
-		let hash = tx.hash();
-
 		// Skip checking the contract if the transaction does not have zero gas price
 		if !tx.gas_price.is_zero() {
 			return Ok(false)
 		}
 
-		let address = client.registry_address(SERVICE_TRANSACTION_CONTRACT_REGISTRY_NAME.to_owned(), BlockId::Latest)
+		self.check_address(client, sender)
+	}
+
+	/// Checks if given address is whitelisted to send service transactions.
+	pub fn check_address<C: CallContract + RegistryInfo + ChainInfo>(&self, client: &C, sender: Address) -> Result<bool, String> {
+		let cache = self.certified_addresses_cache.cache.try_read();
+		if (client.chain_info().best_block_number as u64 == self.certified_addresses_cache.last_block) && cache.is_some() {
+			if let Some(allowed) = cache.unwrap().get(&sender) {
+				return Ok(*allowed);
+			}
+		}
+		let contract_address = client.registry_address(SERVICE_TRANSACTION_CONTRACT_REGISTRY_NAME.to_owned(), BlockId::Latest)
 			.ok_or_else(|| "contract is not configured")?;
+		trace!(target: "txqueue", "Checking service transaction checker contract from {}", sender);
+		let (data, decoder) = service_transaction::functions::certified::call(sender);
+		let value = client.call_contract(BlockId::Latest, contract_address, data)?;
+		decoder.decode(&value).and_then(|allowed| {
+			let cache = self.certified_addresses_cache.cache.try_write();
+			if cache.is_some() {
+				cache.unwrap().insert(sender, allowed);
+			};
+			Ok(allowed)
+		}).map_err(|e| e.to_string())
+	}
 
-		trace!(target: "txqueue", "[{:?}] Checking service transaction checker contract from {}", hash, sender);
+	/// Refresh certified addresses cache
+	pub fn refresh_cache<C: CallContract + RegistryInfo + ChainInfo>(&mut self, client: &C) -> Result<bool, String> {
+		trace!(target: "txqueue", "Refreshing certified addresses cache");
+		let contract_address = client.registry_address(SERVICE_TRANSACTION_CONTRACT_REGISTRY_NAME.to_owned(), BlockId::Latest);
+		if contract_address.is_none() {
+			return Ok(false)
+		}
+		let mut updated_addresses: BTreeMap<Address, bool> = BTreeMap::default();
+		let cache = self.certified_addresses_cache.cache.try_read();
+		if cache.is_some() {
+			for (address, allowed_before) in cache.unwrap().iter() {
+				let (data, decoder) = service_transaction::functions::certified::call(*address);
+				let value = client.call_contract(BlockId::Latest, contract_address.unwrap(), data)?;
+				let allowed = decoder.decode(&value).map_err(|e| e.to_string())?;
+				if *allowed_before != allowed {
+					updated_addresses.insert(*address, allowed);
+				}
+			};
+			let cache = self.certified_addresses_cache.cache.try_write();
+			if cache.is_some() {
+				let mut unwrapped_cache = cache.unwrap();
+				for (address, allowed) in updated_addresses.iter() {
+					unwrapped_cache.insert(*address, *allowed);
+				}
+				self.certified_addresses_cache.last_block = client.chain_info().best_block_number as u64;
```

Comment:
So this normally would need to be under a lock as well, but since your function is `&mut self` (you are modifying only one instance) you don't need it.